move/squid/sources/squid/swap_info.move

```diff
@@ -3,102 +3,107 @@ module squid::swap_info;
 
 public(package) fun new(data: vector<u8>, ctx: &mut TxContext): SwapInfo {
     let swap_data = bcs::new(data).peel_vec_vec_u8();
     SwapInfo {
-        swap_index: 0,
-        estimate_index: 0,
-        status: ESTIMATING,
+        status: SwapStatus::Estimating{index: 0, fallback: false},
         coin_bag: coin_bag::new(ctx),
         swap_data,
     }
 }
 
-public(package) fun get_data_swapping(self: &mut SwapInfo): vector<u8> {
-    let index = self.swap_index;
-    if (index == 0 && self.status == ESTIMATING) {
-        assert!(
-            self.estimate_index == self.swap_data.length(),
-            ENotDoneEstimating,
-        );
-        self.status = SWAPPING;
+public(package) fun get_data_swapping(self: &mut SwapInfo): (vector<u8>, bool) {
+    let (index, fallback) = match(self.status) {
+        SwapStatus::Estimating{index: _, fallback: _} => abort(ENotDoneEstimating),
+        SwapStatus::Swapping{index, fallback} => (index, fallback),
+        SwapStatus::Done => abort(EDoneSwapping),
     };
+
     assert!(index < self.swap_data.length(), EOutOfSwaps);
 
-    self.swap_index = index + 1;
-    if (self.status == SKIP_SWAP) {
-        vector[]
+    self.status = if (index + 1 < self.swap_data.length()) {
+        SwapStatus::Swapping{index: index + 1, fallback}
     } else {
-        assert!(self.status == SWAPPING, ENotSwapping);
-        self.swap_data[index]
-    }
+        SwapStatus::Done
+    };
+
+    (self.swap_data[index], fallback)
 }
 
-public(package) fun get_data_estimating(self: &mut SwapInfo): vector<u8> {
-    let index = self.estimate_index;
-    assert!(index < self.swap_data.length(), EOutOfEstimates);
+public(package) fun get_data_estimating(self: &mut SwapInfo): (vector<u8>, bool) {
+    let (index, fallback) = match(self.status) {
+        SwapStatus::Estimating{index, fallback} => (index, fallback),
+        _ => abort(EDoneEstimating),
+    };
 
-    self.estimate_index = index + 1;
+    assert!(index < self.swap_data.length(), EOutOfEstimates);
 
-    if (self.status == SKIP_SWAP) {
-        vector[]
+    self.status = if (index + 1 < self.swap_data.length()) {
+        SwapStatus::Estimating{index: index + 1, fallback}
     } else {
-        assert!(self.status == ESTIMATING, ENotEstimating);
-        self.swap_data[index]
-    }
+        SwapStatus::Swapping{index: 0, fallback}
+    };
+    
+    (self.swap_data[index], fallback)
 }
 
 public(package) fun coin_bag(self: &mut SwapInfo): &mut CoinBag {
     &mut self.coin_bag
 }
 
-public(package) fun swap_data(self: &SwapInfo, i: u64): vector<u8> {
-    self.swap_data[i]
-}
-
 public(package) fun skip_swap(self: &mut SwapInfo) {
-    self.status = SKIP_SWAP;
+    self.status = match(self.status) {
+        SwapStatus::Estimating{index, fallback} => {
+            assert!(!fallback, EAlreadySkippingSwaps);
+            SwapStatus::Estimating{index, fallback: true}
+        },
+        _ => abort(ENotEstimating),
+    };
```

Comment:
```suggestion
    self.status = match(self.status) {
        SwapStatus::Estimating{ index, fallback: false } => {
            SwapStatus::Estimating{ index, fallback: true }
        },
        SwapStatus::Estimating{ .. } => abort(EAlreadySkippingSwaps),
        _ => abort(ENotEstimating),
    };
```
you can also add match guards such as `SwapStatus::Estimating{ index, fallback } if (fallback)`
for more complex conditions

Reply:
nice